flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendBuilder.java

```diff
@@ -0,0 +1,502 @@
+				this.savepointSnapshotStrategy,
+				writeBatchWrapper,
+				defaultColumnFamilyHandle,
+				nativeMetricMonitor);
+			PriorityQueueSetFactory priorityQueueFactory = initPriorityQueueFactory(backend);
+			backend.setPriorityQueueFactory(priorityQueueFactory);
+
+			if (restorePrepareResult != null) {
+				postBackendInitialization(backend, restorePrepareResult);
+			}
+
+			return backend;
+		} catch (Throwable e) {
+			// Do clean up
+			if (backend != null) {
+				backend.dispose();
+			} else {
+				IOUtils.closeQuietly(writeOptions);
+				IOUtils.closeQuietly(defaultColumnFamilyHandle);
+				IOUtils.closeQuietly(writeBatchWrapper);
+				IOUtils.closeQuietly(nativeMetricMonitor);
+				IOUtils.closeQuietly(cancelStreamRegistry);
+			}
+			try {
+				FileUtils.deleteDirectory(instanceBasePath);
+			} catch (Exception ex) {
+				LOG.warn("Failed to instance base path for RocksDB: " + instanceBasePath, ex);
+			}
+			// Log and rethrow
+			if (e instanceof BackendBuildingException) {
+				throw (BackendBuildingException) e;
+			} else {
+				String errMsg = "Caught unexpected exception.";
+				LOG.error(errMsg, e);
+				throw new BackendBuildingException(errMsg, e);
+			}
+		}
+	}
+
+	private void initializeSnapshotStrategy(CloseableRegistry cancelStreamRegistry,
+											ResourceGuard rocksDBResourceGuard,
+											LinkedHashMap<String, StateColumnFamilyHandle> kvStateInformation,
+											int keyGroupPrefixBytes) {
+		this.savepointSnapshotStrategy = new RocksFullSnapshotStrategy<>(
+			db,
+			rocksDBResourceGuard,
+			keySerializer,
+			kvStateInformation,
+			keyGroupRange,
+			keyGroupPrefixBytes,
+			localRecoveryConfig,
+			cancelStreamRegistry,
+			keyGroupCompressionDecorator);
+		if (enableIncrementalCheckpointing) {
+			// TODO eventually we might want to separate savepoint and snapshot strategy, i.e. having 2 strategies.
+			checkpointSnapshotStrategy = new RocksIncrementalSnapshotStrategy<>(
+				db,
+				rocksDBResourceGuard,
+				keySerializer,
+				kvStateInformation,
+				keyGroupRange,
+				keyGroupPrefixBytes,
+				localRecoveryConfig,
+				cancelStreamRegistry,
+				instanceBasePath,
+				backendUID,
+				materializedSstFiles,
+				lastCompletedCheckpointId,
+				numberOfTransferingThreads);
+		} else {
+			checkpointSnapshotStrategy = savepointSnapshotStrategy;
+		}
+	}
+
+	private PriorityQueueSetFactory initPriorityQueueFactory(RocksDBKeyedStateBackend<K> backend) {
+		PriorityQueueSetFactory priorityQueueFactory;
+		switch (priorityQueueStateType) {
+			case HEAP:
+				priorityQueueFactory = new HeapPriorityQueueSetFactory(keyGroupRange, numberOfKeyGroups, 128);
+				break;
+			case ROCKSDB:
+				priorityQueueFactory = new RocksDBPriorityQueueSetFactory(backend);
+				break;
+			default:
+				throw new IllegalArgumentException("Unknown priority queue state type: " + priorityQueueStateType);
+		}
+		return priorityQueueFactory;
+	}
+
+	private void postBackendInitialization(
```

Comment:
I think this method could be improved a bit. In particular, we should be able to do and prepare most of the things this method does before the `backend` is created, as soon as we have the `db`. This would allow us to do all the incremental restore related things much closer to opening the db, so maybe we can just have two methods to open a db with incremental state and a method to open an empty db. The only part that "requires" a backend is `RocksDBIncrementalRestoreOperation.getOrRegisterColumnFamilyHandle`. But this call looks already strange and we can probably remove the class `RocksDBIncrementalRestoreOperation` anyways. Actually,  I think having the db is enough for that method as well and you can just add meta data to the maps that are later passed inside the backend constructor. This could improve cleanliness and simplicity of the code and it would also represent a proper logical order, first preparing everyting, then creating the backend instead of first creating the backend and then "postprocessing" some internals.